Include/object.h

```diff
@@ -81,12 +81,25 @@ typedef struct _typeobject PyTypeObject;
 /* PyObject_HEAD defines the initial segment of every PyObject. */
 #define PyObject_HEAD                   PyObject ob_base;
 
+/* Immortalization:
+ * This marks the reference count bit that will be used to define immortality.
+ * The GC bit-shifts refcounts left by two, and after that shift it still needs
+ * to be larger than zero, so it's placed after the first three high bits.
+ */
+#define _Py_IMMORTAL_BIT_OFFSET (8 * sizeof(Py_ssize_t) - 4)
+#define _Py_IMMORTAL_BIT (1LL << _Py_IMMORTAL_BIT_OFFSET)
```

Comment:
If we set the initial value to something halfway up to the next bit then compatibility should be less of a problem for older extensions built against the limited API.  We'd use that value in `PyObject_HEAD_IMMORTAL_INIT` and `_Py_SetImmortal()`.  We could still check against `_Py_IMMORTAL_BIT` as-is (i.e.. in `_Py_IsImmortal()`).

See https://github.com/python/cpython/pull/24828/files#diff-2a12f738a77b362d74a65949b58c37f2affcd15ba8b1c979b63bd00223b8a456R64-R69.

Reply:
Probably the main motivation to keep it as a single bit is to preserve space for future features that might use other bits. Who knows, maybe in a few years from now we'll need the 5th and 6th MSB to mark some other state. i.e perhaps in-object tri-color marking to get towards a mark & sweep GC (purely speculation at this point). The main point being that I don't want immortalization to constrain the evolution of the language

Though keeping a strong compatibility story should also for sure be weighed into what we choose for the value here. Do you have some specific examples where setting the value to halfway up the immortal bit would help keep better compatibility? At least internally, we didn't have any compatibility issues using the immortal bit even though we had countless extensions that we didn't initially recompile with the newer API

Reply:
Packing other data into the refcount is an interesting idea.  I don't think we have to worry about the precise initial refcount at this point though.  The bit is still the marker, so it doesn't matter what the value is as long as it has that bit set.  It's also opaque to the API, thus we aren't tying ourselves down.  So I don't see any risk with using an initial value higher than the bit.

On the other hand, setting the higher initial value should help avoid problems with some older extensions.  For example, let's say this PR is merged and someone using 3.11 imports an extension that was compiled against an older Python version using the limited API (ergo stable ABI).  Then that extension was compiled with the old `Py_DECREF()` and `Py_INCREF()`.  Therefore, the refcount for any object it touches is going to fluctuate.  So we already cannot guarantee that the refcount will not change (at least not without excluding any such extensions).  By using an initial refcount higher than the bit we can at least be confident that the "immortal" bit will still be set.

(Older "limited API" extensions like that can cause similar issues via `Py_SET_REFCNT()` but there's no way around that.)

Reply:
Oh I see! Yeah, we were treating extension objects that were incref/decref as "lost" immortalized objects since we were purely focusing on Copy on Writes. But you are right, it's probably better to keep the immortal bit active even when an instance is used in an older extension.

Also, you are right; as long as we keep the bit market as the actual check the original value doesn't really matter.

Let me update the PR!

Reply:
Sorry for not being more clear in the first place. :)